coordination/coordination.go

```diff
@@ -11,4 +16,195 @@ type Client interface {
+//
+// In general, Session API is concurrency-friendly, you can safely access all of its methods concurrently.
+//
+// The client guarantees that sequential calls of the methods are sent to the server in the same order. However, the
+// session client may reorder and suspend some of the requests without violating correctness of the execution. This also
+// applies to the situations when the underlying gRPC stream has been recreated due to network or server issues.
+//
+// The client automatically keeps the underlying gRPC stream alive by sending keep-alive (ping-pong) requests. If the
+// client can no longer consider the session alive, it immediately cancels the session context which also leads to
+// cancellation of contexts of all semaphore leases created by this session.
+type Session interface {
+	// Close closes the coordination service session. It cancels all active requests on the server and notifies every
+	// pending or waiting for response request on the client side. It also cancels the session context and tries to
+	// stop the session gracefully on the server. If the ctx is canceled, this will not wait for the server session to
+	// become stopped and returns immediately with an error. Once this function returns with no error, all subsequent
+	// calls will be noop.
+	Close(ctx context.Context) error
+
+	// Context returns the context of the session. It is canceled when the underlying server session is over or if the
+	// client could not get any successful response from the server before the session timeout (see
+	// options.WithSessionTimeout).
+	Context() context.Context
+
+	// CreateSemaphore creates a new semaphore. This method waits until the server successfully creates a new semaphore
+	// or returns an error.
+	//
+	// This method is not idempotent. If the request has been sent to the server but no reply has been received, it
+	// returns the ErrOperationStatusUnknown error.
+	CreateSemaphore(ctx context.Context, name string, limit uint64, opts ...options.CreateSemaphoreOption) error
+
+	// UpdateSemaphore changes semaphore data. This method waits until the server successfully updates the semaphore or
+	// returns an error.
+	//
+	// This method is idempotent. The client will automatically retry in the case of network or server failure.
+	UpdateSemaphore(ctx context.Context, name string, opts ...options.UpdateSemaphoreOption) error
+
+	// DeleteSemaphore deletes an existing semaphore. This method waits until the server successfully deletes the
+	// semaphore or returns an error.
+	//
+	// This method is not idempotent. If the request has been sent to the server but no reply has been received, it
+	// returns the ErrOperationStatusUnknown error.
+	DeleteSemaphore(ctx context.Context, name string, opts ...options.DeleteSemaphoreOption) error
+
+	// DescribeSemaphore returns the state of the semaphore.
+	//
+	// This method is idempotent. The client will automatically retry in the case of network or server failure.
```

Comment:
I think at least Describe with watches cannot be idempotent, but you don't seem to support them. Are you planning to add watches as a separate method?

Reply:
At the moment, there are no watches in the client, so it should be idempotent. I've been thinking about supporting this feature, however I'm not sure what the API will look like.